### DISCUSSION TITLE
Connection pool

### DISCUSSION THREAD

**Discussion by ZeN220** [pika/pika#1425] [Q&A]
I haven't found any implementation of the connection pool in the library. Is it due to some specific reason or it just hasn't been implemented yet?

I would like to see examples of its implementation (in async), if they are available

**Comment by lukebakken** [comment]
Nobody has implemented a connection pool using Pika. I think, in general, most users find one connection to be sufficient. If you wish to use more than one then connection-per-thread will work.

If you need assistance with your code, share what you have tried (as a git repository) and I can review it.

**Reply by ZeN220** [reply]
my application is a chatbot (into telegram). it uses rabbitmq quite heavily and sends a lot of requests. accordingly, if the RPS is high enough, one connection may not be enough

in DBMS this problem is solved by connection pooling. i.e. there are N already open connections, which are fetched for each query and returned to the pool after the query is processed.

how would you implement connection pooling in pika?

**Comment by lukebakken** [comment]
> accordingly, if the RPS is high enough, one connection may not be enough

Maybe, but you won't know unless you actually run benchmarks. And yes, I know what a connection pool is 😉 

> how would you implement connection pooling in pika?

If, through benchmarks and real-world use, you find that a single connection is limiting your application, you have the following options:

* Run more application processes, each with its own connection. I.e. "pool" at the process level.
* Run a connection per Python thread. Be sure that calls to Pika are thread-safe by **only** using the `connection.add_callback_threadsafe` method to schedule their execution.

If you want an actual opaque connection pool you will have to implement a higher-level messaging client on top of Pika that abstracts the low-level AMQP concepts like connections, channels, etc. Good examples of projects like this are NServiceBus and MassTransit, both in the .NET world.
